libraries/Microsoft.Bot.Builder.Dialogs/SkillDialog.cs

```diff
@@ -197,5 +208,66 @@ private async Task<Activity> SendToSkillAsync(ITurnContext context, Activity act
 
             return eocActivity;
         }
+
+        private async Task<bool> InterceptOAuthCardsAsync(ITurnContext turnContext, Activity activity, string connectionName, CancellationToken cancellationToken)
+        {
+            if (activity?.Attachments != null && !string.IsNullOrWhiteSpace(connectionName))
+            {
+                foreach (var attachment in activity.Attachments.Where(a => a?.ContentType == OAuthCard.ContentType))
+                {
+                    var oauthCard = ((JObject)attachment.Content).ToObject<OAuthCard>();
+                    if (!string.IsNullOrWhiteSpace(oauthCard.TokenExchangeResource?.Uri))
+                    {
+                        var tokenExchangeProvider = turnContext.Adapter as IExtendedUserTokenProvider;
+                        if (tokenExchangeProvider == null)
+                        {
+                            return false;
+                        }
+
+                        TokenResponse result;
+                        try
+                        {
+                            result = await tokenExchangeProvider.ExchangeTokenAsync(
+                               turnContext,
+                               connectionName,
+                               turnContext.Activity.From.Id,
+                               new TokenExchangeRequest(oauthCard.TokenExchangeResource.Uri)).ConfigureAwait(false);
+                        }
+                        catch
```

Comment:
what exception is this catching? can we make it more specific?

Reply:
And is it worth logging this exception? Future maintainers of this code have ZERO way of knowing / understanding why this catch is there ignoring the exception. So there are 2 layers here: 1) we'd love to understand why we do this and provided that it makes sense to all of us, 2) Please document your decision with comments.

Reply:
The idea is that if token exchange fails for any kind of exception (server goofing up, some casting issue), just show the oauthcard to the user to unblock them.
I can add documentation, but do you want it logged as well? (It would have been logged by the adapter hopefully)

Reply:
Agreed, we can't do a blanket catch; shouldn't the prompt know if it should retry or not?